src/lib/pubkey/classic_mceliece/cmce_field_ordering.cpp

```diff
@@ -0,0 +1,322 @@
+ */
+secure_vector<uint16_t> generate_control_bits_internal(const secure_vector<uint16_t>& pi) {
+   auto n = pi.size();
+   size_t m = 1;
+
+   while(size_t(1) << m < n) {
+      m += 1;
+   }
+
+   BOTAN_ASSERT_NOMSG(size_t(1) << m == n);
+
+   if(m == 1) {
+      return secure_vector<uint16_t>({pi.at(0)});
+   }
+   secure_vector<uint16_t> p(n);
+   for(size_t x = 0; x < n; ++x) {
+      p.at(x) = pi.at(x ^ 1);
+   }
+   secure_vector<uint16_t> q(n);
+   for(size_t x = 0; x < n; ++x) {
+      q.at(x) = pi.at(x) ^ 1;
+   }
+
+   secure_vector<uint16_t> range_n(n);
+   std::iota(range_n.begin(), range_n.end(), 0);
+   auto piinv = composeinv(range_n, pi);
+
+   simultaneous_composeinv(p, q);
+
+   secure_vector<uint16_t> c(n);
+   for(uint16_t x = 0; size_t(x) < n; ++x) {
+      c.at(x) = CMCE_CT::min(x, p.at(x));
+   }
+
+   simultaneous_composeinv(p, q);
+
+   for(size_t i = 1; i < m - 1; ++i) {
+      auto cp = composeinv(c, q);
+      simultaneous_composeinv(p, q);
+      for(size_t x = 0; x < n; ++x) {
+         c.at(x) = CMCE_CT::min(c.at(x), cp.at(x));
+      }
+   }
+
+   secure_vector<uint16_t> f(n / 2);
+   for(size_t j = 0; j < n / 2; ++j) {
+      f.at(j) = c.at(2 * j) % 2;
+   }
+
+   secure_vector<uint16_t> big_f(n);
+   for(uint16_t x = 0; size_t(x) < n; ++x) {
+      big_f.at(x) = x ^ f.at(x / 2);
+   }
+
+   auto fpi = composeinv(big_f, piinv);
+
+   secure_vector<uint16_t> l(n / 2);
+   for(size_t k = 0; k < n / 2; ++k) {
+      l.at(k) = fpi.at(2 * k) % 2;
+   }
+
+   secure_vector<uint16_t> big_l(n);
+   for(uint16_t y = 0; size_t(y) < n; ++y) {
+      big_l.at(y) = y ^ l.at(y / 2);
+   }
+
+   auto big_m = composeinv(fpi, big_l);
+
+   secure_vector<uint16_t> subm0(n / 2);
+   secure_vector<uint16_t> subm1(n / 2);
+   for(size_t j = 0; j < n / 2; ++j) {
+      subm0.at(j) = big_m.at(2 * j) / 2;
+      subm1.at(j) = big_m.at(2 * j + 1) / 2;
+   }
+
+   auto subz0 = generate_control_bits_internal(subm0);
+   auto subz1 = generate_control_bits_internal(subm1);
+
+   secure_vector<uint16_t> z(subz0.size() + subz1.size());
+   for(size_t j = 0; j < subz0.size(); ++j) {
+      z.at(2 * j) = subz0.at(j);
+      z.at(2 * j + 1) = subz1.at(j);
+   }
+
+   return concat(f, z, l);
+}
+
+}  // anonymous namespace
+
+std::optional<Classic_McEliece_Field_Ordering> Classic_McEliece_Field_Ordering::create_field_ordering(
+   const Classic_McEliece_Parameters& params, std::span<const uint8_t> random_bits) {
+   BOTAN_ARG_CHECK(random_bits.size() == (params.sigma2() * params.q()) / 8, "Wrong random bits size");
+
+   secure_vector<uint32_t> a;  // contains a_0, a_1, ...
+   for(size_t i = 0; i < params.q(); ++i) {
+      a.push_back(load_le<uint32_t>(random_bits.data(), i));  // Utilizes sigma2 = 32
+   }
+
+   auto pi = create_pi(a);
+
+   for(size_t i = 1; i < params.q(); ++i) {
+      if(a.at(i - 1) == a.at(i)) {  // Check for duplicate elements utilizing sorting
+         return std::nullopt;       // Abort (Step 2)
+      }
+   }
```

Comment:
Today I learnt (by means of CoPilot) that C++20 actually has a function for that ([std::adjacent_find](https://en.cppreference.com/w/cpp/algorithm/adjacent_find)):

```suggestion
   // Check for duplicate elements (Step 1)
   if(std::adjacent_find(pi.begin(), pi.end()) != pi.end()) {
      return std::nullopt;  // Abort (Step 2)
   };
```

Reply:
Nice! `a` instead of `pi` must be checked, though. I'll apply it!